Added nodes routing table

bortoletti-giorgia opened the pull request:
In the interface Routing Table there are a lot of red and green lines only because I have changed the order of methods.

bortoletti-giorgia committed:
Merge pull request #48 from RicDeZen/feature/distributedNetwork

Feature/distributed network in feature/distributedNetwork-RoutingTable

bortoletti-giorgia committed:
Revert "Merge branch 'feature/distributedNetwork-RoutingTable' into feature/distributedNetwork"

This reverts commit 9c5ef103a673c4739c4dc7fbae47d3a5d438bc74, reversing
changes made to af5d27fc7e03a86cfd44a2123c2940f97ead538e.

bortoletti-giorgia committed:
Implementation of Peers Routing Table

bortoletti-giorgia committed:
Updated methods getClosest and changed PeersRoutingTable in NodesRoutingTable

Now NodesRT works with a generic Node. There is only NodesRoutingTable with PeerNode, missed tests with ResourceNode.

bortoletti-giorgia committed:
Updated method of comparison between nodes

Now to find distance in NodesRoutingTable it's used position of most significant bit at 1 of BitSet distance XOR.

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/BinarySet.java:
```diff
@@ -63,11 +63,14 @@ public BinarySet getDistance(@NonNull BinarySet set) {
      */
-    public int getDistanceInteger(@NonNull BinarySet set) {
-        return new BigInteger(getDistance(set).getKey().toByteArray()).intValue();
+    public int getFirstPositionOfOne() {
```
Not very useful, you can do this directly with bitset.length()

https://docs.oracle.com/javase/7/docs/api/java/util/BitSet.html#length()

bortoletti-giorgia replied:
bitset.length() returns the lenght but I need the most significant bit at 1.

turcato-niccolo replied:
Returns the "logical size" of this BitSet: the index of the highest set bit in the BitSet plus one. Returns zero if the BitSet contains no set bits.

bortoletti-giorgia replied:
Now I have changed the implementation.

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NodesRoutingTable.java:
```diff
@@ -0,0 +1,154 @@
+     */
+    @Override
+    public boolean contains(Node node) {
+        if(node == null || bucketsTable.size() == 0)
```
`bucketsTable.isEmpty` instead of `bucketsTable.size() == 0`?

bortoletti-giorgia replied:
It's true, thank you.

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NodesRoutingTable.java:
```diff
@@ -0,0 +1,154 @@
+     */
+    @Override
+    public Node getClosest(Node node) {
+        if(!node.equals(nodeOwner)) {
```
returning null isn't quite correct in this case, because the closest node to the owner node in the routing table is (if present) in the last buck (bucket number N) or in the N-1, or N-2, etc...
And it will be useful to call this method for the owner Node if a Node is asked of the closest node (or the k closest nodes) of the network to himself.

bortoletti-giorgia replied:
> > Which things? I can't find them.
> > FindBucket was deleted and also Big Integer.
> 
> Did you receive them now? My fault. they remained pending from yesterday, I hadn't noticed.

Now I'll correct all.

bortoletti-giorgia replied:
Already corrected in the last commit but thank you for your review.

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NodesRoutingTable.java:
```diff
@@ -0,0 +1,154 @@
+    public int getLocation(Node node) {
+        if(node == null)
+            return -1;
+        Object distanceObject = nodeOwner.getDistance(node);
```
You could avoid this situation defining nodeOwner as Node<BinarySet>, or either casting ((Node<BinarySet>)nodeOwner) or eventually ((PeerNode)nodeOwner). With this last solution, you should check in the constructor if the node is an exemplar of PeerNode.

bortoletti-giorgia replied:
nodeOwner is already a Node but with a generic type so I must have the certainty of this generic type is BinarySet. Otherwise I should declare nodeOwner as Node<BinaryKey> and only in the KBucket.
Do you think this replacement could be the best solution or do I always check that if the distance is a Binary Set?

turcato-niccolo replied:
Well, it does not make a lot of sense to work with generic Nodes in the routing table since it's based on binary addresses. Maybe it should accept Node<BinarySet> only?

bortoletti-giorgia replied:
I think it's fair that the routing table only accepts Node of BinarySet, now I fix it.

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NodesRoutingTable.java:
```diff
@@ -0,0 +1,154 @@
+
+    /**
+     * @param i index of the bucket in buckets container
+     * @return the bucket at index i
```
This specification is not complete, if i isn't in its bound this method returns null.

bortoletti-giorgia replied:
It's true, thank you.

turcato-niccolo commented on distributednetworkmanager/src/main/java/com/gruppo1/distributednetworkmanager/NodesRoutingTable.java:
```diff
@@ -0,0 +1,154 @@
+     */
+    @Override
+    public Node[] getKClosest(Node node) {
+        if(!node.equals(nodeOwner)) {
```
Same as before, Note that the closest node to the owner node in the routing table is (if present) in the last buck (bucket number N) or in the N-1, or N-2, etc...
You could return the bucket of index N-1, but it would contain always 1 node (at best), it would be better to return a bunch of buckets starting from the N-1 index (descending)

bortoletti-giorgia replied:
Already corrected in the last commit but thank you for your review.

bortoletti-giorgia committed:
Merge branch 'feature/distributedNetwork' into feature/distributedNetwork-RoutingTable

bortoletti-giorgia committed:
Added NodesRoutingTableTest with mock test

In this test distance between two nodes is implementing with method getDistance of PeerNode

turcato-niccolo commented:
Nice job with getKClosest and getClosest.

Some of the things I pointed out were not solved nor you responded. Didn't you agree with me on those?

bortoletti-giorgia commented:
Which things? I can't find them.
FindBucket was deleted and also Big Integer.

turcato-niccolo commented:
> Which things? I can't find them.
> FindBucket was deleted and also Big Integer.

Did you receive them now? My fault they remained pending prom yesterday, I hadn't noticed.

bortoletti-giorgia committed:
Resolved some specifics after Turcato's review

bortoletti-giorgia committed:
Now RoutingTable and KBucket are made of Node<BinarySet>

I have replaced Node with Node<BinarySet> because nodes are based on binary addresses.

turcato-niccolo reviewed (approved):
Nice, thanks👍

bortoletti-giorgia committed:
Changed implementation of BinarySet.getFirstPositionOfOne()

turcato-niccolo reviewed (approved):
Thank you👍
